engine/schema/src/com/cloud/host/dao/HostDao.java

```diff
@@ -103,4 +95,13 @@
     List<HostVO> listByType(Type type);
 
     HostVO findByIp(String ip);
+
+    /**
+     * This method will look for a host that is connected to the storage pool where the volume of the Snapshot is stored.
+     * <ul>
+     * <li>If the storage pool found for the volume of the snapshotId has more than a host, we will choose one randomly;
+     * <li>If no host is found, we throw a runtime exception
+     * </ul>
+     */
+    HostVO findHostToOperateOnSnapshot(long snapshotId);
```

Comment:
@rafaelweingartner,  Command only goes to SSVM when only secondary storage is involved so no role of storage pool here so better to think of some other way to find host.

Reply:
I did not understand this, can you explain what you mean with some more details?

Reply:
For unmanaged storage snapshots goes on secondary storage. Secondary storage is accessible to all hosts in zone. So any random host in a zone will do. Once the snapshot is created for unmanaged storage it's independent of volume and storage pool. I believe managed store is zone wide so it is available to all and any random host in zone. Also it's handling is bit different. Confirm behaviour with @mike-tutkowski.

Reply:
@anshul1886  Any host at all?
I mean, if I create a snapshot of a volume that is being used in a VM in a XenServer host. Can I use a Vmware ESXi host to transform this snapshot into a volume to be used by a XenServer VM?

Reply:
Currently managed storage is just supported at the zone level. I do have code in mind that would allow it to be confined to cluster scope, though. You make a good point that such a configuration (since snapshots for managed storage are stored on primary storage) would make the use of snapshots outside of that cluster somewhat complicated.

Reply:
@mike-tutkowski, following @anshul1886 I changed this code to look for any host in the zone that is of the same hypervisor type of the cluster where the original volume of the snapshot resides on. 

So, what do you guys think? let´s use only hosts of the cluster where the volume of the snapshot has been taken resides on? For me, it seems an easy way to cope with these different scenarios that we have.

Reply:
One workaround I already have in place, though, is the following (currently only supported on XenServer): You can pass in a parameter when taking the volume snapshot that tells us to put the volume snapshot on secondary storage instead of keeping it on primary storage. If, however, you have already created the snapshot on primary storage, then you'd need to create a volume from it and then snapshot that volume to secondary storage.

Reply:
So, this requires that we use a host that has access to the primary storage where the snapshot resides?

Reply:
On XenServer, making use of a volume snapshot that resides on primary storage needs to leverage a XenServer host that supports UUID resigning. UUID resigning is available as a patch for XenServer (@syed created it in 2016). When the CSMS connects with a XenServer host, it checks to see if this patch is installed. The results of the check are stored in cloud.host_details (the supportsResign property). If any host in the cluster supports UUID resigning, then the "convenience" property, also called supportsResign, in cloud.cluster_details is set to true. That being the case, we need to find one of these hosts in the zone when leveraging a managed snapshot on XenServer.

Reply:
"So, this requires that we use a host that has access to the primary storage where the snapshot resides?"

Correct, but currently we only support zone-wide managed storage, so all hosts in the zone should have access to that primary storage.

Reply:
@mike-tutkowski I will take your comments into consideration. This new requirement will change things a bit, but at least we will make a comprehensive solution for snapshots in XenServer deployments

As soon as I have an update, I will notify you guys